components/ca-oauth2-client.html

```diff
@@ -0,0 +1,211 @@
+         * The following events are fired by this element:
+         *
+         * <dl>
+         * <dt>ca-oauth2-authorized</dt>
+         * <dd>Fired when authorization has been successful, before the credentials are provided to <ca-auth>.
+         * Call {@code Event.preventDefault()} to stop this happening. Event detail is the OAuth2 token response.</dd>
+         * <dt>ca-oauth2-unauthorized</dt>
+         * <dd>Fired when authorization has been unsuccessful. Event detail is the error message.</dd>
+         * </dl>
+         *
+         * @see https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2
+         */
+        class CaOAuth2ClientElement extends HTMLElement {
+            /**
+             * observedAttributes - get a list of attributes that should be considered in the callback.
+             * @return {Array} list of attributes that should be considered in the callback.
+             */
+            static get observedAttributes() {
+                return ['href', 'type', 'client-id', 'client-secret', 'username', 'password'];
+            }
+
+            /**
+             * Called when any attribute is changed on the element, but is only interested in the {@link CaOAuth2ClientElement.observedAttributes}.
+             * Will trigger the authorization request if the element has sufficient attributes set to initiate a flow.
+             * @param {string} attrName - the name of the attribute to have changed
+             * @param {string} oldVal - the old value of the attribute
+             * @param {string} newVal - the new value of the attribute
+             * @returns {undefined} nothing
+             */
+            attributeChangedCallback(attrName, oldVal, newVal) {
+                if (CaOAuth2ClientElement.observedAttributes.includes(attrName)) {
+                    if (this.ready()) {
+                        this.requestToken();
+                    }
+                }
+            }
+
+            /**
+             * @return {string} the URI of the authorization server's endpoint (token endpoint for password grant)
+             */
+            get href() {
+                return this.getAttribute('href');
+            }
+
+            /**
+             * @param {string} value, the URI of the authorization server's endpoint (token endpoint for password grant)
+             */
+            set href(value) {
+                this.setAttribute('href', value);
+            }
+
+            /**
+             * @return {string} the grant type, defaults to 'password'
+             */
+            get type() {
+                return this.getAttribute('type') || 'password';
+            }
+
+            /**
+             * @param {string} value, the grant type - one of 'password', 'code', 'implicit' or 'client'.
+             */
+            set type(value) {
+                this.setAttribute('type', value);
```

Comment:
the grant type - one of 'password', 'code', 'implicit' or 'client'. -> Could this not check the value against an accept list of values and then default to password if nothing else matches?

Reply:
Good idea; well actually the JSDoc is somewhat forward looking, for now it'll have to be 'password' regardless of what you set. But I think it should check the value and then leave it alone if it doesn't match the list, rather than set it to password.